src/middlewared/middlewared/plugins/boot_/environments.py

```diff
@@ -0,0 +1,185 @@
+import datetime
+import errno
+import json
+import os
+import subprocess
+
+from middlewared.api import api_method
+from middlewared.api.current import (
+    BootEnvironmentActivateArgs,
+    BootEnvironmentActivateResult,
+    BootEnvironmentCloneArgs,
+    BootEnvironmentCloneResult,
+    BootEnvironmentDestroyArgs,
+    BootEnvironmentDestroyResult,
+    BootEnvironmentEntry,
+    BootEnvironmentKeepArgs,
+    BootEnvironmentKeepResult,
+)
+from middlewared.service import filterable_api_method, Service, private
+from middlewared.service_exception import CallError, ValidationError
+from middlewared.utils import filter_list
+from middlewared.utils.size import format_size
+
+
+def run_zectl_cmd(cmd):
+    cmd.insert(0, "zectl")
+    try:
+        cp = subprocess.run(
+            cmd,
+            capture_output=True,
+            text=True,
+            check=True,
+        )
+    except subprocess.CalledProcessError as cpe:
+        err = f"STDOUT: {cpe.stdout.strip()!r} STDERR: {cpe.stderr.strip()!r}"
+        raise CallError(f"Unexpected error: {err}")
+    else:
+        return cp.returncode == 0
+
+
+class BootEnvironmentService(Service):
+    class Config:
+        namespace = "boot.environment"
+        entry = BootEnvironmentEntry
+        cli_private = True
+
+    @private
+    def zfs_get_props(self):
+        zget = "zfs get -r -d 1"
+        zget += " truenas:kernel_version,zectl:keep,used,creation"
+        bp_name = self.middleware.call_sync("boot.pool_name")
+        root = f"{bp_name}/ROOT"
+        json_flags = "-j --json-int"  # json format, raw numbers
+        try:
+            cp = subprocess.run(
```

Comment:
Should we start working on new py-libzfs-free plugin now? Generalizing code looks like a perfect first method for new ZFS service.

Reply:
It's not a bad idea, however, we are going to start working on `py-libzfs2` (I've got a NEP that I can share with you). In `py-libzfs2`, we will be working with upstream to make `libzfs*` thread-safe and re-entrant. This would mean we could use native libzfs bindings inside coroutines safely. Anyways, we could potentially start writing our own separate package for doing things via zfs/zpool using `-j` flag but for right now, i'll keep this as-is.